marker_lints/src/not_using_has_span_trait.rs

```diff
@@ -0,0 +1,44 @@
+use marker_api::prelude::*;
+
+marker_api::declare_lint! {
+    /// ### What it does
+    /// Suggests using `impl HasSpan` for functions that need to take `Span` as
+    /// a parameter to make them more ergonomic.
+    NOT_USING_HAS_SPAN_TRAIT,
+    Warn,
+}
+
+pub(crate) fn check_item<'ast>(cx: &'ast MarkerContext<'ast>, item: ItemKind<'ast>) {
+    let ItemKind::Fn(func) = item else { return };
+
+    for param in func.params() {
+        let ast::TyKind::Path(path) = strip_references(param.ty()) else {
+            continue;
+        };
+
+        let ident = path
+            .path()
+            .segments()
+            .last()
+            .unwrap_or_else(|| panic!("BUG: path must have at least one segment, but got: {path:#?}"))
+            .ident();
+
+        if ident != "Span" {
+            continue;
+        }
+
+        cx.emit_lint(
+            NOT_USING_HAS_SPAN_TRAIT,
+            func,
+            "use impl HasSpan instead of Span for more flexibility",
+        )
+        .span(param.ty().span());
+    }
+}
+
+fn strip_references(ty: ast::TyKind<'_>) -> ast::TyKind<'_> {
+    match ty {
+        ast::TyKind::Ref(ref_ty) => strip_references(ref_ty.inner_ty()),
+        _ => ty,
+    }
```

Comment:
I wonder if this function could be moved to the public API of `TyKind`.

Reply:
it would definitely be a good function to have. [peel_refs](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.peel_refs) from rustc might be a good example as well

Reply:
Cool 🤙, I added the same `peel_refs` method to `ast::TyKind` and `sem::TyKind`.

Reply:
I've noticed that there is no information about the parentheses around the type in `TyKind`s. For example both `u32`, `(u32)`, and `((u32))` are represented the same in the AST. This is in contrast to `syn`, where it has a [`Paren`](https://docs.rs/syn/latest/syn/struct.TypeParen.html) type kind that wraps the inner type in parens explicitly.

I added two test cases to `print_ty.rs` to show that there is no info about that. Is this by design? I don't think this is critical, and in fact the `Paren` type kind in `syn` is often annoying and I have to write the same `strip_ty_group` function that removes the parentheses and an invisible [`Group`](https://docs.rs/syn/latest/syn/struct.TypeGroup.html). It's really hard to understand what `Group` is, but I suppose that's related to invisible groupping that is added by the compiler when a macro is invoked in type position. For example `&dyn macro_call!()` is implicitly grouped when expanded like `&dyn (expanded_ty_here)` and the parentheses I used here aren't visible.

Reply:
> Is this by design?

Yes, such information is usually not represented in ASTs, as they are semantically the same. You'll see the same thing for binary operators. `(1 + 2) * 3` the brackets are not included in the final AST. The tree structure preserves the execution order.

This is part of what the abstract from "Abstract Syntax Tree" is about. It tries to unify these syntactical differences into one uniform view, at the loss of information. If a user wants, they can still use the span, to see if there are parentheses, but that's not necessarily recommended.

Clippy has a few lints, which check for the syntax. But they're always a bit weird, as we use Clippy as a formatter with semantic information. [clippy::semicolon_if_nothing_returned](https://rust-lang.github.io/rust-clippy/master/index.html#/semicolon_if_nothing_returned) is an example, where we need to check the written syntax as well for some heuristics, if the semicolon should be suggested.

> So I'm not really sure marker needs to expose any parens info at all. Stripping them automatically simplifies life, but maybe there could be lints that want that info? I'm not sure.

In Clippy we have lived happy without this information, and hacked around with spans, where needed. If we wanted to provide a way to completely reconstruct the original source code, we should include parenthesis. Rustfmt and syn, for instance, require this information. For Marker's case, I don't believe that this should be one of our goals. It would be nice to have a function to print the AST as it would be written (Similar how rustc has this for hir, with the "Show HIR" option on the playground). But adding parenthesis nodes, would add a lot of edge cases to linting, with very little benefit :)